File: 1833-maximum-ice-cream-bars/solution.cpp
#include <vector>
#include <algorithm>

class Solution {
public:
    int maxIceCream(std::vector<int>& costs, int coins) {
        // Find the maximum cost to determine the size of the frequency array
        int max_cost = 0;
        for (int cost : costs) {
            max_cost = std::max(max_cost, cost);
        }

        // Create frequency array
        std::vector<int> freq(max_cost + 1, 0);
        for (int cost : costs) {
            freq[cost]++;
        }

        int count = 0;
        // Iterate through costs in ascending order
        for (int cost = 1; cost <= max_cost; ++cost) {
            if (freq[cost] > 0) {
                // Determine how many bars of this price we can afford
                long long can_buy = std::min((long long)freq[cost], (long long)coins / cost);
                
                count += (int)can_buy;
                coins -= (int)(can_buy * cost);
                
                // If we cannot afford anymore of the current price, we stop
                if (coins < cost) break;
            }
        }
        
        return count;
    }
};

File: 1833-maximum-ice-cream-bars/solution.py
class Solution:
    def maxIceCream(self, costs: list[int], coins: int) -> int:
        # Determine the range for counting sort
        max_cost = max(costs)
        freq = [0] * (max_cost + 1)
        
        # Fill the frequency array
        for cost in costs:
            freq[cost] += 1
            
        count = 0
        # Iterate through frequencies to buy cheapest bars first
        for cost in range(1, max_cost + 1):
            if freq[cost] > 0:
                # Calculate how many of this specific price we can afford
                num_bars = min(freq[cost], coins // cost)
                
                count += num_bars
                coins -= (num_bars * cost)
                
                # If we have less coins than the current cost, we can't buy anymore
                if coins < cost:
                    break
                    
        return count